Refactoring

Ambratolm opened the pull request:
Changes so far:
- All code elements (except localized strings properties names and some comments) renamed to English.
- All long recurring methods shortened.
- API full rework complete.

Commit:
Minor adjustments in API utilities.

Commit:
Turned all forms names & main form controls names into english.

Commit:
Shortened LanguagesResourceManager GetString metheod.

Widely used global API method: LanguagesResourceManager.GetString(name, culture) shortened to: GetString(name) and all occurences of the method across the project replaced with the new one.

Commit:
Shortened MessageBox Show metheod.

 Widely used: MessageBox.Show(text, caption, buttons, icon, defaultButton, options); shortened to: QuickMessageBox.ShowX(text) where X represents type of the message (information, question, warning, or error) and all occurences of the method across the project replaced with the new one.

Commit:
API divided and organized into more particular classes and sub-folders.

- API global class divided into more particular classes: App, Database, and Language.
- API folder organized into sub-folders: Core and UI Elements.
- All calls to API properties and methods across the project are renamed accordingly.

Commit:
Assets folder creation logic wrapped into API method.

AXeL-dev commented:
I can see that you kinda liked the app :grin: , this app is really old, like maybe 5 or 6 years ago haha, but anyway, i'm glad you enjoy playing with the code & refactoring it.

Let me know when you finish refactoring so i merge all the changes at once! (this way you can still push new commits to this PR).

Commit:
Shortened Language Checking Expression.

- (Language.GetCurrentLanguage() == "ar") shortened to: (Language.IsArabic)
- All occurences of the expression replaced with the new one.

Commit:
Removed UI logic from API database FetchAllTables method.

API database method FetchAllTables decoupled from UI ToolStripStatuslabel parameter. now the ToolStripStatuslabel is set within the main form logic.

Commit:
File name DateTime formatting methods replaced with simple alternative.

Statement (DateTime.Now.ToString().Replace("/", "-").Replace(":", "-").Replace(" ", "_")) replaced with (DateTime.Now.ToString("yyyy-MM-dd_hh-mm-ss")) in all occurences in project.

Commit:
Backup WinRar compression logic wrapped into API class.

Commit:
Language setting additional logic wrapped in API method.

Now changing current culture, setting message boxes reading direction, and saving language to database is done through one method only.

Commit:
Multi-language support system improved.

- Added available languages resources files detection.
- Added support for more than two languages.
- Added language menu-item sub-items auto-generation from available languages resources files (Adding a new language resource file will auto-add that language to application with no need for further setup which gives better support for adding more new translations).
- Added japanese language resource file.

Commit:
Added strongly-typed access support to language resource file strings.

- Added invariant general language resource file for the generation of strongly-typed class that allows access to localized strings through class properties instead of the non-typed GetString method of ResourceManager.
- All GetString calls in project replaced accordinly.

Commit:
LanguageChanged Event moved to API.

- As it's not really related to Form_Main, LanguageChanged moved to API class, and now to subscribe to the event, no more need to call an instance of Form_Main every time.
- All event subscription logic in project replaced accordinly.

Commit:
Updated forms single instances checking logic.

- Removed API Forms Tracking Boolean Properties.
- Now single form opening is checked through form instances properties.

Commit:
All controls names and texts renamed to english.

Commit:
Database API reworked (Work in progress).

- API code shortened using data access layer classes.
- Now DataTables can be accessed directly through API properties using a wrapper data access class.
- All occurences in project replaced accordingly.
- Some bugs need to be fixed (To do for next commit).

Commit:
Database turned to english (Work in progres).

- Database tables and columns renamed to english.
- All references to databae tables and columns names replaced accordingly.
- Data folder (database file and assets folder) added to project tree.
- Added post build event to copy data folder to output folder.
- Fixed some bugs (More to fix in next commit)

Commit:
Fixed bugs.

Runtime exceptions fixed, now the app can run :D
- The table "Language" name is a reserved word in access database, it caused errors in the DataAdapter object trying to run the generated SQL commands. Renaming it to "CurrentLanguage" fixed the errors.
- Renamed some forgotten references to the old names of the renamed database fields that caused exceptions when loading some forms.

Commit:
Job and Help forms refactored.

[Forms refactoring work in progression]
- Job and Help forms refactored.
Some API methods added or updated due to form refactoring needs.
- Added GetRows and HasRows methods in TableAccessor class of Database API.
- Expression in refactored forms replaced accordingly.

Commit:
Refactored AddCustomer form and Added Assets API.

[Forms refactoring work in progression]
- AddCustomer form refactored.
- Added Assets API class for better control over the creation of assets folders and files.
- API path properties (Database & Assets paths) placed in their respective API classes (Database & Assets).
- Removed AppFolderPath property becuase it's not needed as the paths are natively resolved in a relative way (to the current app executable folder path).
- References affected by changes in project replaced accordingly.

Commit:
Imrpoved Assets API.

Commit:
Customer forms refactored (Except EditCustomer form).

[Forms refactoring work in progression]
- AddCustomer form refactored.
- CustomersList form refactored.
- SearchCustomer form refactored.
- Job and Help forms refactored.
- Added DataTableNavigator utility class.
- Added DataGridViewColumnFormat utility class.

Commit:
EditCustomer form refactored.

[Forms refactoring work in progression]
EditCustomer form sucks >:-(

AXeL-dev commented:
Dude, that's a lot of work/refactoring! :astonished: "ma3yitish? xD"

Ambratolm commented:
> Dude, that's a lot of work/refactoring!  "ma3yitish? xD"

Hello, I was working on this during holidays, it was fun 💃. but the holidays are done, and I couldn't find time to finish it. well, I think commits up to `8d0e4a0` are ok ✔️ (the app runs without errors), but changes beyond it have some new bugs 🐞 as I remember 😆 I hope it gets some attention 🌹 in free times. I know it's kind of an old forgotten project 👵, but it's not about the project, it's about the message 🃏.

"🌏👽🚀 kawkab zo7al hal min tar7ib!?"
